### ISSUE TITLE
How to throw a validation error without a 500 status code?

### ISSUE THREAD

**Issue by good-idea** [graphql/express-graphql#422]
I'm trying to display a simple "Email and password do not match" error on a login form, but am having trouble and not sure if where I've gone wrong is on the server or in my React app. It seems like it has something to do with how I am using express-graphql, but I'm not sure.

Here's the resolver on the server:

    // Schema definition:
    
    type Mutation {
       loginViewer(credentials: AUTH_CREDENTIALS): SignInPayload!
    }
    
    type SignInPayload {
       token: String
       expires: Int
       requiresReset: Boolean
    }
    
    // Mapping the mutation to the resolver:
    
    Mutation: {
       loginViewer: loginViewerMutation,
    },

   
    // Resolver:

    const loginViewerMutation = async (obj, args) => {
       const { credentials } = args
       const user = await User.findOne({ email })
       if (!user) throw new GraphQLError('Email and password do not match')
       const matches = await user.comparePassword(password)
       if (!matches) throw new GraphQLError('Email and password do not match')
       return createJWT(user)
    }

Then, the mutation as it's called within my React Login component:

    const mutation = gql`
       mutation LoginViewer($password: String!, $email: String!) {
          loginViewer(credentials: { email: $email, password: $password }) {
             token
             expires
             requiresReset
          }
       }
    `
    
    export class Login extends React.Component {
       handleLogin = (credentials) => {
          this.props
             .mutate({ variables: { ...credentials } })
             .then(() => {
                // ...
             })
             .catch((error) => {
                console.log(error.message)
                console.log(error.graphQLErrors)
                // ...
             })
       }
    
       render() {
          // ...
       }
    }
    
    export default graqphql(mutation)(Login)
      

Everything works as expected when I supply the correct information. When I don't, the error that is caught doesn't contain the GraphQLErrors.

I'm using the apollo-link-error middleware with the default setup: https://www.apollographql.com/docs/link/links/error.html

My console looks like this:

[![enter image description here][1]][1]

Now, the error message ("Email and password do not match") *is* being sent with the response. But, it's always alongside a 500 network error, and react-apollo discards its graphQLErrors array when there's a network error present.

Where might I be going wrong here? It doesn't seem right to be returning a 500 Internal Server Error for a user-facing error --- the server behaving exactly as I want it to. Is there a different way that I should be throwing this error? 

  [1]: https://i.stack.imgur.com/uPWL0.png

**Comment by IvanGoncharov** [comment]
@good-idea 
**TL; DR;** Remove exclamation mark at this line: 
`loginViewer(credentials: AUTH_CREDENTIALS): SignInPayload!`

Personal opinion: If you expect some field to return an error during normal execution you shouldn't make it Non-Null.

---------------------------
It's expected behavior since your top-level field (`loginViewer`) defined as non-null. That means when error bubble up it GraphQL engine has no option except make `data` field equals to null:
> Since Non-Null type fields cannot be null, field errors are propagated to be handled by the parent field. If the parent field may be null then it resolves to null, otherwise if it is a Non-Null type, the field error is further propagated to it’s parent field.

> If all fields from the root of the request to the source of the error return Non-Null types, then the  "data" entry in the response should be null.

http://facebook.github.io/graphql/draft/#sec-Errors-and-Non-Nullability

If `data` is null that means the entire request is failed and `express-graphql` response with 500.
